src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2027,31 +2027,50 @@ private NullableFlowState GetDefaultState(ref LocalState state, int slot)
             var variable = _variables[slot];
             var symbol = variable.Symbol;
 
-            switch (symbol.Kind)
+            switch (symbol)
             {
-                case SymbolKind.Local:
+                case LocalSymbol local:
                     {
-                        var local = (LocalSymbol)symbol;
                         if (!_variables.TryGetType(local, out TypeWithAnnotations localType))
                         {
                             localType = local.TypeWithAnnotations;
                         }
                         return localType.ToTypeWithState().State;
                     }
-                case SymbolKind.Parameter:
+                case ParameterSymbol parameter:
                     {
-                        var parameter = (ParameterSymbol)symbol;
                         if (!_variables.TryGetType(parameter, out TypeWithAnnotations parameterType))
                         {
                             parameterType = parameter.TypeWithAnnotations;
                         }
                         return GetParameterState(parameterType, parameter.FlowAnalysisAnnotations).State;
                     }
-                case SymbolKind.Field:
-                case SymbolKind.Property:
-                case SymbolKind.Event:
+                case PropertySymbol { Name: WellKnownMemberNames.ValuePropertyName } property when
+                        variable.ContainingSlot is > 0 and var containingSlot &&
+                        property.ContainingType.IsWellKnownTypeIUnion() &&
+                        _variables[containingSlot].Symbol.GetTypeOrReturnType().Type is NamedTypeSymbol { IsUnionTypeNoUseSiteDiagnostics: true, UnionCaseTypes: not [] } unionType &&
+                        Binder.GetUnionTypeValuePropertyNoUseSiteDiagnostics(unionType) == (object)property:
+                    {
+                        // For union types where none of the case types are nullable, the default state for Value is "not null" rather than "maybe null".
+                        var result = NullableFlowState.NotNull;
+
+                        foreach (var ctor in unionType.InstanceConstructors)
```

Comment:
Could we not immediately break if any of the cases are `MaybeDefault`? And if we can, is there any way to detect up front that generics aren't involved, and therefore break immediately if any of the cases are `MaybeNull`?